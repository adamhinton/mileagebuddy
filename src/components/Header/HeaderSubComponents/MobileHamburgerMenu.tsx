// _______________________________________________________________________________
// This is the hamburger dropdown icon for Header.tsx
// Note, this is just the icon, nothing else
// This only appears on mobile; on larger screen sizes the menu is always visible.
// _______________________________________________________________________________

import ThemeSwitch from "@/components/ThemeSwitch";

type MobileHamburgerMenuProps = {
	isMobileMenuOpen: boolean;
	setIsMobileMenuOpen: (isOpen: boolean) => void;
};

const MobileHamburgerMenu = (props: MobileHamburgerMenuProps) => {
	const { isMobileMenuOpen, setIsMobileMenuOpen } = props;

	return (
		<div className="sm:hidden flex items-center space-x-3">
			{/* ThemeSwitch is the dark/light mode toggle icon */}
			<ThemeSwitch />
			<button
				onClick={() => setIsMobileMenuOpen(!isMobileMenuOpen)}
				className="p-2 rounded-md text-neutral-text hover:bg-background-highlight"
				aria-label="Toggle mobile menu"
			>
				{isMobileMenuOpen ? (
					// X icon for close
					<svg
						xmlns="http://www.w3.org/2000/svg"
						className="h-6 w-6"
						fill="none"
						viewBox="0 0 24 24"
						stroke="currentColor"
					>
						<path
							strokeLinecap="round"
							strokeLinejoin="round"
							strokeWidth={2}
							d="M6 18L18 6M6 6l12 12"
						/>
					</svg>
				) : (
					// Hamburger icon
					<svg
						xmlns="http://www.w3.org/2000/svg"
						className="h-6 w-6"
						fill="none"
						viewBox="0 0 24 24"
						stroke="currentColor"
					>
						<path
							strokeLinecap="round"
							strokeLinejoin="round"
							strokeWidth={2}
							d="M4 6h16M4 12h16M4 18h16"
						/>
					</svg>
				)}
			</button>
		</div>
	);
};

export default MobileHamburgerMenu;
