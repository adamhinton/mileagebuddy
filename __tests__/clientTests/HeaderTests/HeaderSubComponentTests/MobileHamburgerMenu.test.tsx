// ________________________________________________________________________
// This is (obviously) tests for MobileHamburgerMenu.tsx
// The component is pretty simple, so not much to test here
// ________________________________________________________________________

import MobileHamburgerMenu from "@/components/Header/HeaderSubComponents/MobileHamburgerMenu";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";

// mock setIsMobileMenuOpen

describe("MobileHamburgerMenu.tsx", () => {
	it("Renders without errors with menu closed", () => {
		render(
			<MobileHamburgerMenu
				isMobileMenuOpen={false}
				setIsMobileMenuOpen={jest.fn()}
			/>
		);
	});

	describe("Component Display States", () => {
		it("displays hamburger icon when menu is closed", () => {
			const { container } = render(
				<MobileHamburgerMenu
					isMobileMenuOpen={false}
					setIsMobileMenuOpen={jest.fn()}
				/>
			);

			// Check that the hamburger icon path is present
			const hamburgerPath = container.querySelector(
				'path[d="M4 6h16M4 12h16M4 18h16"]'
			);
			expect(hamburgerPath).toBeInTheDocument();

			// Make sure X icon is not present
			const xPath = container.querySelector('path[d="M6 18L18 6M6 6l12 12"]');
			expect(xPath).not.toBeInTheDocument();
		});

		it("displays X icon when menu is open", () => {
			const { container } = render(
				<MobileHamburgerMenu
					isMobileMenuOpen={true}
					setIsMobileMenuOpen={jest.fn()}
				/>
			);

			// Check that the X icon path is present
			const xPath = container.querySelector('path[d="M6 18L18 6M6 6l12 12"]');
			expect(xPath).toBeInTheDocument();

			// Make sure hamburger icon is not present
			const hamburgerPath = container.querySelector(
				'path[d="M4 6h16M4 12h16M4 18h16"]'
			);
			expect(hamburgerPath).not.toBeInTheDocument();
		});
	});
	it("Renders without errors with menu open", () => {
		render(
			<MobileHamburgerMenu
				isMobileMenuOpen={true}
				setIsMobileMenuOpen={jest.fn()}
			/>
		);
	});

	it("Calls setIsMobileMenuOpen when clicked", () => {
		const setIsMobileMenuOpenMock = jest.fn();
		const { getByLabelText } = render(
			<MobileHamburgerMenu
				isMobileMenuOpen={false}
				setIsMobileMenuOpen={setIsMobileMenuOpenMock}
			/>
		);

		const button = getByLabelText("Toggle mobile menu");
		button.click();

		expect(setIsMobileMenuOpenMock).toHaveBeenCalledWith(true);
	});

	it("Calls setIsMobileMenuOpen when clicked again", () => {
		const setIsMobileMenuOpenMock = jest.fn();
		const { getByLabelText } = render(
			<MobileHamburgerMenu
				isMobileMenuOpen={true}
				setIsMobileMenuOpen={setIsMobileMenuOpenMock}
			/>
		);

		const button = getByLabelText("Toggle mobile menu");
		button.click();

		expect(setIsMobileMenuOpenMock).toHaveBeenCalledWith(false);
	});

	describe("MobileHamburgerMenu Screenshots", () => {
		it("matches snapshot when closed", () => {
			const { asFragment } = render(
				<MobileHamburgerMenu
					isMobileMenuOpen={false}
					setIsMobileMenuOpen={jest.fn()}
				/>
			);
			expect(asFragment()).toMatchSnapshot();
		});

		it("matches snapshot when open", () => {
			const { asFragment } = render(
				<MobileHamburgerMenu
					isMobileMenuOpen={true}
					setIsMobileMenuOpen={jest.fn()}
				/>
			);
			expect(asFragment()).toMatchSnapshot();
		});
	});
});
