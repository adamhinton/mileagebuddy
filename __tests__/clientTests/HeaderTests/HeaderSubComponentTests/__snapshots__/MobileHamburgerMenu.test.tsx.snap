// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileHamburgerMenu.tsx MobileHamburgerMenu Screenshots matches snapshot when closed 1`] = `
<DocumentFragment>
  <div
    class="sm:hidden flex items-center space-x-3"
  >
    <button
      aria-label="Toggle mobile menu"
      class="p-2 rounded-md text-neutral-text hover:bg-background-highlight"
    >
      <svg
        class="h-6 w-6"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 6h16M4 12h16M4 18h16"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`MobileHamburgerMenu.tsx MobileHamburgerMenu Screenshots matches snapshot when open 1`] = `
<DocumentFragment>
  <div
    class="sm:hidden flex items-center space-x-3"
  >
    <button
      aria-label="Toggle mobile menu"
      class="p-2 rounded-md text-neutral-text hover:bg-background-highlight"
    >
      <svg
        class="h-6 w-6"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 18L18 6M6 6l12 12"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
